File: update-nfr.py
import os
import requests
import json

print("Getting NFR Reactions")

comments = 'https://api.github.com/repos/phalcon/cphalcon/issues/14608/comments'
headers = {
    'Accept': 'application/vnd.github.squirrel-girl-preview+json',
    'User-Agent': 'Phalcon Agent'
}

token = os.environ.get('GITHUB_TOKEN')
if token:
    headers['Authorization'] = f"Bearer {token}"

result = {}
page = 1
while True:
    url = f"{comments}?per_page=100&page={page}"

    response = requests.get(url, headers=headers, timeout=30)
    response.raise_for_status()
    content = response.text

    print(f"Got page {page}")
    data = json.loads(content)
    if not isinstance(data, list):
        break

    for comment in data:
        id = comment.get('id', '')
        url = comment.get('html_url', '')
        body = comment.get('body', '')
        reactions = comment.get('reactions', {})
        plusone = reactions.get('+1', 0)

        body = body.split('\n')[0].replace('\r', '').replace('\r\n', '')
        body = body.replace('\\', '\\\\').replace('|', '\\|').replace('[', '\\[').replace(']', '\\]')
        plusone = f"{int(plusone):03}"
        result[f"{plusone}-{id}"] = {
            'reaction': plusone,
            'body': f"[{body}]({url})"
        }

    if len(data) < 100:
        break
    page += 1

print("Sorting Results")
sorted_result = dict(sorted(result.items(), key=lambda item: item[0], reverse=True))

print("Creating Content")
output = "# New Feature Request List\n"
output += "- - -\n\n"
output += "| Votes  | Description             |\n"
output += "|--------|-------------------------|\n"
for item in sorted_result.values():
    output += f"| {item['reaction']} | {item['body']} |\n"

output += "\n"

file_name = 'docs/new-feature-request-list.md'
with open(file_name, 'w') as file:
    file.write(output)

File: scripts/update-nfr.mjs
/**
 * Refresh the "New Feature Request List" page of every version from the
 * reactions on the comments of cphalcon issue #14608.
 *
 *   pnpm run update-nfr          (GITHUB_TOKEN raises the API rate limit)
 *
 * Each comment becomes one table row: its 👍 count and its first line,
 * linked to the comment. Rows are sorted by votes, then by comment id.
 */
import { existsSync, writeFileSync } from "node:fs";
import { fileURLToPath, pathToFileURL } from "node:url";
import { dirname, join } from "node:path";
import { versions } from "../src/versions.generated.mjs";

const COMMENTS_URL = "https://api.github.com/repos/phalcon/cphalcon/issues/14608/comments";

/** Escape a comment line for a Markdown table cell inside MDX. */
export function escapeCell(text) {
  return text
    .replace(/\\/g, "\\\\")
    .replace(/\|/g, "\\|")
    .replace(/\[/g, "\\[")
    .replace(/\]/g, "\\]")
    .replace(/[{}]/g, (brace) => "\\" + brace)
    .replace(/</g, "&lt;");
}

/**
 * @param {Array<{ id: number, html_url: string, body: string, reactions?: { "+1"?: number } }>} comments
 * @returns {string} The MDX page.
 */
export function formatList(comments) {
  const rows = comments
    .map((comment) => {
      const votes = Number(comment.reactions?.["+1"] ?? 0);
      const line = (comment.body ?? "").split("\n")[0].replace(/\r/g, "");
      return { votes, id: comment.id, cell: `[${escapeCell(line)}](${comment.html_url})` };
    })
    .sort((a, b) => b.votes - a.votes || b.id - a.id);
  return [
    "---",
    'title: "New Feature Request List"',
    "---",
    "",
    "| Votes  | Description             |",
    "|--------|-------------------------|",
    ...rows.map((row) => `| ${String(row.votes).padStart(3, "0")} | ${row.cell} |`),
    "",
  ].join("\n");
}

export async function fetchComments(fetchImpl = fetch) {
  const headers = {
    Accept: "application/vnd.github.squirrel-girl-preview+json",
    "User-Agent": "Phalcon Agent",
  };
  if (process.env.GITHUB_TOKEN) headers.Authorization = `Bearer ${process.env.GITHUB_TOKEN}`;
  const comments = [];
  for (let page = 1; ; page++) {
    const response = await fetchImpl(`${COMMENTS_URL}?per_page=100&page=${page}`, { headers });
    if (!response.ok) throw new Error(`GitHub API: ${response.status} ${response.statusText}`);
    const data = await response.json();
    if (!Array.isArray(data)) break;
    comments.push(...data);
    console.log(`Got page ${page}`);
    if (data.length < 100) break;
  }
  return comments;
}

async function main() {
  const root = join(dirname(fileURLToPath(import.meta.url)), "..");
  const comments = await fetchComments();
  const page = formatList(comments);
  // Every version that has the page gets the same, current list.
  for (const version of versions) {
    const target = join(root, "src", "content", `docs-${version}`, "new-feature-request-list.mdx");
    if (!existsSync(target)) continue;
    writeFileSync(target, page, "utf8");
    console.log(`Wrote ${comments.length} requests to ${target}`);
  }
}

if (process.argv[1] && import.meta.url === pathToFileURL(process.argv[1]).href) {
  await main();
}
